File: GraphColoredModel.cs
using System;
using System.Collections.Generic;

public class GraphColoringSimulatedAnnealing
{
    private int[] colors; // ノードの色
    private int[,] adjacencyMatrix; // 隣接行列
    private int numNodes; // ノードの数
    private int numColors; // 色の数
    private double temperature; // 温度
    private double coolingRate; // 冷却率

    public GraphColoringSimulatedAnnealing(int[,] adjacencyMatrix, int numColors, double initialTemperature, double coolingRate)
    {
        this.adjacencyMatrix = adjacencyMatrix;
        this.numNodes = adjacencyMatrix.GetLength(0);
        this.numColors = numColors;
        this.temperature = initialTemperature;
        this.coolingRate = coolingRate;
    }

    // グラフ彩色問題を解く
    public int[] Solve()
    {
        // ノードの色を初期化
        colors = new int[numNodes];

        // 初期解をランダムに生成
        Random random = new Random();
        for (int i = 0; i < numNodes; i++)
        {
            colors[i] = random.Next(numColors);
        }

        // 焼きなまし法を実行
        while (temperature > 0.1)
        {
            for (int i = 0; i < 100; i++) // 100回の反復ステップ
            {
                // ランダムなノードを選択
                int randomNode = random.Next(numNodes);

                // ノードの色を変更してエネルギーを計算
                int oldColor = colors[randomNode];
                int newColor = random.Next(numColors);
                colors[randomNode] = newColor;
                int deltaEnergy = CalculateDeltaEnergy(randomNode, oldColor, newColor);

                // 変更を受け入れるか判定
                if (deltaEnergy < 0 || random.NextDouble() < Math.Exp(-deltaEnergy / temperature))
                {
                    // 変更を受け入れる
                    continue;
                }
                else
                {
                    // 変更を破棄して元の色に戻す
                    colors[randomNode] = oldColor;
                }
            }

            // 温度を下げる
            temperature *= coolingRate;
        }

        return colors;
    }

    // ノードの色変更によるエネルギーの変化を計算
    private int CalculateDeltaEnergy(int node, int oldColor, int newColor)
    {
        int deltaEnergy = 0;
        for (int i = 0; i < numNodes; i++)
        {
            if (adjacencyMatrix[node, i] == 1 && i != node)
            {
                if (colors[i] == oldColor && colors[i] != newColor)
                {
                    deltaEnergy++;
                }
                else if (colors[i] == newColor && colors[i] != oldColor)
                {
                    deltaEnergy--;
                }
            }
        }
        return deltaEnergy;
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[,] adjacencyMatrix = new int[,]
        {
            { 0, 1, 1, 0 },
            { 1, 0, 1, 1 },
            { 1, 1, 0, 1 },
            { 0, 1, 1, 0 }
        };

        int numColors = 3;
        double initialTemperature = 100.0;
        double coolingRate = 0.95;

        GraphColoringSimulatedAnnealing solver = new GraphColoringSimulatedAnnealing(adjacencyMatrix, numColors, initialTemperature, coolingRate);
        int[] colors = solver.Solve();

        Console.WriteLine("Node Colors:");
        for (int i = 0; i < colors.Length; i++)
        {
            Console.WriteLine($"Node {i + 1}: Color {colors[i] + 1}");
        }
    }
}

File: graph_colored_model.py
import numpy as np
import random
import math


class GraphColoringSimulatedAnnealing:
    def __init__(self, adjacency_matrix, num_colors, initial_temperature, cooling_rate):
        self.adjacency_matrix = adjacency_matrix
        self.num_nodes = adjacency_matrix.shape[0]
        self.num_colors = num_colors
        self.temperature = initial_temperature
        self.cooling_rate = cooling_rate

    def solve(self):
        # ノードの色を初期化
        colors = np.random.randint(
            low=0, high=self.num_colors, size=self.num_nodes)

        # 焼きなまし法を実行
        while self.temperature > 0.1:
            for _ in range(100):  # 100回の反復ステップ
                # ランダムなノードを選択
                random_node = random.randint(0, self.num_nodes - 1)

                # ノードの色を変更してエネルギーを計算
                old_color = colors[random_node]
                new_color = random.randint(0, self.num_colors - 1)
                delta_energy = self.calculate_delta_energy(
                    colors, random_node, old_color, new_color)

                # 変更を受け入れるか判定
                if delta_energy < 0 or random.random() < math.exp(delta_energy / self.temperature):
                    # 変更を受け入れる
                    colors[random_node] = new_color
                else:
                    # 変更を破棄して元の色に戻す
                    colors[random_node] = old_color

            # 温度を下げる
            self.temperature *= self.cooling_rate

        return colors

    def calculate_delta_energy(self, colors, node, old_color, new_color):
        delta_energy = 0
        for i in range(self.num_nodes):
            if self.adjacency_matrix[node, i] == 1 and i != node:
                if colors[i] == old_color and colors[i] != new_color:
                    delta_energy += 1
                elif colors[i] == new_color and colors[i] != old_color:
                    delta_energy -= 1
        return delta_energy


adjacency_matrix = np.array([[0, 1, 1, 0],
                             [1, 0, 1, 1],
                             [1, 1, 0, 1],
                             [0, 1, 1, 0]])

num_colors = 4
initial_temperature = 100.0
cooling_rate = 0.95

solver = GraphColoringSimulatedAnnealing(
    adjacency_matrix, num_colors, initial_temperature, cooling_rate)
colors = solver.solve()

print("Node Colors:")
for i in range(len(colors)):
    print(f"Node {i + 1}: Color {colors[i] + 1}")
